### ISSUE TITLE
Is there a reason Laravel doesn't provide Response factory from the controller?

### ISSUE THREAD

**Issue by crnkovic** [laravel/ideas#1280]
I was browsing through the framework's issues/PRs trying to find out is there a specific reason why doesn't Laravel provide some general response "helper" methods in a base controller class, such as `view`, `json`, `response` or `redirect`, so that user generating a JSON response can call `return $this->json()` or respond with a view with `return $this->view()`, kinda like what Symfony provides with `$this->render()`.

I know it's a simple PR, inject Response and View factories and delegate, or resolve right from the helper methods, however I was wondering if there were some issues with that or if there was a reason that Taylor decided not to put it in a framework. I think this would help out developers that are having a lot of responses on their apps and are tired of doing `new Response`, `new JsonResponse` or even worse, `response()->json()`. Also, having `json` method in their base controller class, the users can format their JSON response to their liking. Every controller method should have a response, either view, or JSON, or redirect and this would compliment that by easier response generation.

**Comment by Patryk27** [comment]
I believe the reason is that all those methods would clutter controllers - why add method after method, whereas you can just use `return response()->json();` and call it a day?

Remember that each verb used as method name in the base controller is a verb user **cannot** use anymore - that is not problem in ZF1 or Yii, for example (where each controller's method is named in `xxxAction` / `actionXxx` fashion), but it is in Laravel.

Also, you can just do `return [ ... ];` and Laravel will automatically translate it into JSON.
Also x2, what's wrong with `response()->json()`? It's a good-ol' factory pattern.

**Comment by crnkovic** [comment]
Yeah I guess you make a valid point that these (or any) methods would clutter a controller and make some heavy breaking changes. However there could be `viewFactory`, `responseFactory` properties instead of methods. There are many solutions.

Also, if you take out a `response`, `redirect` and `view` helper functions, a lot of beginners would have no idea how to return a response from a controller, or create a view. That bugs me. I don't like the fact that even the official docs use literally a "helper" function in something built into the core. Notice the irony: helper function for the core functionality. The docs say nothing about `View::make`, or `new JsonResponse` (yes you can return a string or an array, but in terms of kinda object-oriented thinking, you're returning a response from a controller in a form of Response object, not an array). I just don't want this framework and its developers to be too dependent on helper functions that should be used as actual helpers, not do to some core functionality, like render a view. That's why I thought some controller methods should be a good solution, as rendering the view and creating a response instance are primarily used in a controller and very rarely anywhere else.

**Comment by mfn** [comment]
> The documentation is meant for beginners and thus shows the most basic way to return data from controllers - I see no point in telling beginners

To chime in, I wasn't aware of such a rule.

Sure, there are gaps in the docs now and then. After re-reading https://github.com/laravel/ideas/issues/1280#issuecomment-413791812 I felt that this could be solved via education, i.e. the docs.

OTOH: I can understand if people object that the docs get "too crowded". I can't tell were to draw the line.

Personally I think the docs should teach "concepts", and blindly using the helper is not a concept. The concept to teach would be the reader to understand exactly what happens with the returned stuff. There, I said it 😄
